### PR TITLE
Implement `shiftedGreensFunction` in `FFTOpenPoissonSolver` + symmetric Dirichlet test

### PR DESCRIPTION
**Pull request by aliemen**
This pull request adds support for convolving with a shifted Green's function in the FFT open boundary Poisson solver, enabling the method of images for Dirichlet boundary conditions needed in OPALX. It introduces both the implementation and a test for this feature.

## New shifted Green's function support

- Added the `shiftedGreensFunction(const Vector<double, Dim>& shift)` method to `FFTOpenPoissonSolver`, which fills the cached FFT Green's function with one shifted in real space by a user-specified vector. 
- Implemented the `shiftedGreensFunction` method in `FFTOpenPoissonSolver.hpp`, more or less copies/extents `greensFunction` already present in the class. The method is restricted to the HOCKNEY algorithm (and throws an exception if the wrong algorithm is chosen).

### Image charges/Dirichlet example

Added a new integration test, `TestShiftedGreensFunction`, to the CMake test suite, which validates the shifted Green's function implementation by comparing the numerical result (including image correction) against the analytical image-dipole solution for a Gaussian charge near a Dirichlet plane. The test includes detailed diagnostics and passes if both the Dirichlet condition and L2 error are within expected tolerances. 

Note that the "zero plane" is evaluated at the grid point nearest to "0", so the "error" being 4% is exactly as expected. This test only works single rank for now, since flipping a multi-rank field would require a lot more code. 

### Shifted evaluation

An easier test is simply calculating the potential of a bunch located at the origin at a different position. I made such a test (not commited here, since it's just for visualization), where I create a uniform sphere with radius `1e-3` at the origin `(0,0,0)` with `12pC` and calculate the its field with a `1.5e-3` shift in `z` direction. With 64 grid points, the result looks like this:

N64_shifted_greens_z_profile
N64_shifted_greens_yz_sum_overlay

The complete potential is the analytical solution while e.g. the orange line in the 2D plot is the shifted solution of the solver. As one can see, it matches as expected. This should show that at least the shifted Green's function implementation works as expected. The rest of the necessary implementation will either be part of OPALX. Or there will be another PR with a "field flip" helper. 

I will talk about it in detail in today's IPPL meeting.

### CODE REVIEW

**Review by aaadelmann — APPROVED** [review]

File: src/PoissonSolvers/FFTOpenPoissonSolver.hpp [hunk 1]
```diff
@@ -2187,4 +2187,96 @@ namespace ippl {
         }
         ippl::Comm->freeAllBuffers();
     };
+
+    ////////////////////////////////////////////////////////////////////////
+    // Shifted Green's function: fills grn_mr with G(r - shift) on the doubled
+    // grid and caches FFT into grntr_m. Kokkos-parallel, mirrors the HOCKNEY
+    // greensFunction() branch in structure.
```
Comment (s-mayani) [inline src/PoissonSolvers/FFTOpenPoissonSolver.hpp]: "branch" --> method ?

File: src/PoissonSolvers/FFTOpenPoissonSolver.hpp [hunk 2]
```diff
@@ -2187,4 +2187,96 @@ namespace ippl {
...
+        Kokkos::parallel_for(
+            "Shifted Green's function", grn_mr.getFieldRangePolicy(),
+            KOKKOS_LAMBDA(const int i, const int j, const int k) {
+                // local -> global indices
+                const int ig[3] = {i + ldom[0].first() - nghost,
+                                   j + ldom[1].first() - nghost,
+                                   k + ldom[2].first() - nghost};
+
+                // Half-wrap each axis so FFT cyclic convention gives symmetric
+                // physical offsets around the origin:
```
Comment (s-mayani) [inline src/PoissonSolvers/FFTOpenPoissonSolver.hpp]: I am not sure I understand this comment. This is done according to the Hockney method, and has nothing to do with the shift right?

File: src/PoissonSolvers/FFTOpenPoissonSolver.hpp [hunk 3]
```diff
@@ -2187,4 +2187,96 @@ namespace ippl {
...
+
+                // Sign convention matches greensFunction() (HOCKNEY): grn_mr stores -G0
+                // so that solve()'s `rho2tr_m = -rho2tr_m * grntr_m` yields +conv(rho,G0)
+                // where G0(r) = 1/(4 pi |r|).
+                const bool nearSing = (rsq < regThresh);
+                const scalar_type r = Kokkos::sqrt(rsq + nearSing * regThresh);
+                view(i, j, k)       = -1.0 / (4.0 * pi * r);
+            });
+
+        // Fourier-space cache for convolution.
```
Comment (s-mayani) [inline src/PoissonSolvers/FFTOpenPoissonSolver.hpp]: Maybe this comment should be more explicit:
"Store Fourier transform of shifted Green's function for convolution in solve()"

**Review by s-mayani — CHANGES_REQUESTED** [review]

File: src/PoissonSolvers/FFTOpenPoissonSolver.hpp (hunk 2 above)
Comment (aliemen) [reply src/PoissonSolvers/FFTOpenPoissonSolver.hpp]: True it's a bit confusing. The half-wrap is Hockney and the per-axis shift is the Green's shift. I'll simplify the comment.

File: src/PoissonSolvers/FFTOpenPoissonSolver.hpp (hunk 1 above)
Comment (aliemen) [reply src/PoissonSolvers/FFTOpenPoissonSolver.hpp]: Done.

File: src/PoissonSolvers/FFTOpenPoissonSolver.hpp (hunk 3 above)
Comment (aliemen) [reply src/PoissonSolvers/FFTOpenPoissonSolver.hpp]: That's indeed better, thanks

**Review by s-mayani — APPROVED** [review]
